File: 1325.py
from collections import deque
import sys

# TIMEOUT!

n, m = map(int, sys.stdin.readline().split())

edge = []

for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())
    edge.append((a, b))

def bfs(v):
    queue = deque([v])

    visited = [0] * (n+1)
    visited[v] = 1
    pcnum = 0
    
    while queue:
        q = queue.popleft()
        pcnum += 1

        for i in edge:
            if i[1] == q and visited[i[0]] == 0:
                queue.append(i[0])
                visited[i[0]] == 1
    
    return pcnum

max = 0
answer = []

for start in range(1, n+1):
    result = bfs(start)

    if max < result:
        max = result
        answer = [start]
    
    elif max == result:
        answer.append(start)

print(" ".join(map(str, answer)))

File: 1325.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int n, m;
vector<vector<int>> edges;  // Adjacency list representation of the graph

int bfs(int v) {
    queue<int> queue;
    queue.push(v);

    vector<bool> visited(n + 1, false);  // Visited flag for each node
    visited[v] = true;
    int componentSize = 0;

    while (!queue.empty()) {
        int q = queue.front();
        queue.pop();
        componentSize++;

        for (int neighbor : edges[q]) {
            if (!visited[neighbor]) {
                queue.push(neighbor);
                visited[neighbor] = true;
            }
        }
    }

    return componentSize;
}

int main() {
    cin >> n >> m;

    edges.resize(n + 1);  // Resize adjacency list for n nodes (1-based indexing)

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        // edges[a].push_back(b);  // Add edge (a, b) to adjacency list
        edges[b].push_back(a);  // Add edge for undirected graph
    }

    int maxComponentSize = 0;
    vector<int> largestComponents;

    for (int start = 1; start <= n; start++) {
        if (!edges[start].empty()) { // Skip already visited nodes (empty adjacency list)
            int componentSize = bfs(start);
            // cout << "result: " << componentSize << endl;
            if (componentSize > maxComponentSize) {
                maxComponentSize = componentSize;
                largestComponents.clear();
                largestComponents.push_back(start);
            } else if (componentSize == maxComponentSize) {
                largestComponents.push_back(start);
            }
        }
    }

    // cout << maxComponentSize;
    if (!largestComponents.empty()) {
        for (int node : largestComponents) {
            cout << node << " ";
        }
    }
    cout << endl;

    return 0;
}
